Fix TECS init when switching into a altitude controlled mode only once in air

sfuhrer opened the pull request:
### Solved Problem
Fixes https://github.com/PX4/PX4-Autopilot/issues/22487 resp. https://discuss.px4.io/t/cube-o-px4-v1-14-0-erroneous-pitch-roll-setpoint-in-autonomous-mode/35557/11

The issue is only noticeable when the vehicle is switched into an altitude controlled mode when already in air, and there is no airspeed sensor connected. Then the airspeed state is 0, which leads to a division by 0 while calculating the pitch setpoint in TECS.  And TECS initializes wrongly and constrains the INF only to a large value of 5rad.
It is NOT observable if at any time before takeoff the flight mode was already in an altitude controlled mode for a bit, then TECS is correctly initialized by the time it is actually needed.

### Solution
ac33e4eca11ed is the real fix. Prior to this commit, the `_control_params`, eg. `_control_param.pitch_max` , are already used in `TECS::initialize() `but not initialized with the actual settings on the vehicle. As the default for the pitch min/max is 5rad the effect is then clearly noticeable.
e177e9a07b is patching the airspeed-less handling, which is further robustified with 4a8f346d and 09648310. Without these changes, the .tas state is at 0 which then leads to a division by 0, and due to the wrong pitch limits it doesn't get constrained properly.

The other commits are not directly related to the crash, but proposals for mini clean ups and further guards. 


### Changelog Entry
For release notes:
```
Bugfix: Correct TECS initialization when switching into an altitude controlled mode only once in air
```

### Alternatives


### Test coverage
SITL testing. The issue is reproducible in SITL. As the SITL though inits in Position mode, there is a small hack needed to not have it run TECS::update() before switching into Stabilized mode: do not run it within 10s of boot -->  `if (_local_pos_sub.update(&_local_pos) && _local_pos.timestamp > 10_s) `. Further one has to disable the airspeed sensor, I did that by commenting out the publishing of the simulated differential_pressure topic in SimulatorMavlink. 


With current main:
![image](https://github.com/PX4/PX4-Autopilot/assets/26798987/93a6dc0f-ddc4-4c30-b2fb-e092eb57f8df)


With this PR: 
![image](https://github.com/PX4/PX4-Autopilot/assets/26798987/7ffc9904-4e2a-446b-9a83-7a3f65a50a2d)

sfuhrer commented:
@KonradRudin in f57ae5fbd4d21d581759b105736bec1249e1a201 and following I propose to completely remove the manual TECS resetting from the fw_position_controller, and instead do the initialization only at one place anymore: internally in the `TECS::update()` when the dt is too large. This removes a lot of (seemingly) unnecessary complexity without a downside that I found so far.

KonradRudin commented on src/lib/tecs/TECS.cpp:
```diff
@@ -68,9 +68,15 @@ void TECSAirspeedFilter::update(const float dt, const Input &input, const Param
 		return;
 	}
 
+	if (!airspeed_sensor_available) {
+		_airspeed_state.speed = param.equivalent_airspeed_trim;
+		_airspeed_state.speed_rate = 0.f;
+		return;
+	}
+
```
Do we really want this? If you have a airspeed sensor failure mid air, this will result in a jump in the airspeed filter. (although i guess it won't be a high jump). I think you did this to get rid of the initialization at 0 mid air, but this could be done as well in the TECS::initialize function as well? I am not strongly against your solution just want to raise this for discussion and you could do it the other way as well.

sfuhrer replied:
Fair point. Let's do that: 87f73c3714a1667860a61acb522b3d100aaf1150

KonradRudin commented on src/lib/tecs/TECS.cpp:
```diff
@@ -412,6 +418,7 @@ void TECSControl::_calcPitchControl(float dt, const Input &input, const Specific
 	const float pitch_increment = dt * param.vert_accel_limit / math::max(input.tas, FLT_EPSILON);
 	_pitch_setpoint = constrain(pitch_setpoint, _pitch_setpoint - pitch_increment,
 				    _pitch_setpoint + pitch_increment);
+	_pitch_setpoint = constrain(_pitch_setpoint, param.pitch_min, param.pitch_max);
```
This is not really needed, as pitch_increment above is already constrained.

sfuhrer replied:
It is if you want that a changed pitch limit has an effect immediately after changing it. Otherwise it has to go through the increment filter (as seen in the crash that happened).

KonradRudin commented on src/lib/tecs/TECS.cpp:
```diff
@@ -627,8 +641,26 @@ void TECSControl::resetIntegrals()
+
 void TECS::initialize(const float altitude, const float altitude_rate, const float equivalent_airspeed,
-		      const float eas_to_tas)
+		      float eas_to_tas, float throttle_min, float throttle_setpoint_max,
+		      float throttle_trim, float pitch_limit_min, float pitch_limit_max, float target_climbrate,
+		      float target_sinkrate)
```
I think this is just an old remnant of something you tried? Those are not used in the function since it is now only used internally and the initControlParam is set beforehand.

sfuhrer replied:
0e0aee1766082766be83f5a0c11feffd97ee492c

KonradRudin commented on src/modules/fw_pos_control/FixedwingPositionControl.cpp:
```diff
@@ -2535,6 +2522,7 @@ FixedwingPositionControl::tecs_update_pitch_throttle(const float control_interva
 	if (_vehicle_status.is_vtol) {
 		if (_vehicle_status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING || _vehicle_status.in_transition_mode) {
 			_tecs_is_running = false;
+			return;
```
Is this ok? When you transition, it still is rotary wing until it finished transitioning, right? So the code below is not executed and this ramp up is not properly executed (although i would like to get rid of this anyways).

sfuhrer replied:
Yes correct. I'll move the return.
And see https://github.com/PX4/PX4-Autopilot/pull/22510 - there I propose to remove the whole block.

sfuhrer replied:
Fix: f1588db1caf6dbe6895ebd6d20ee84ef87aa98e4

KonradRudin commented on src/lib/tecs/TECS.cpp:
```diff
@@ -627,8 +641,26 @@ void TECSControl::resetIntegrals()
+		      float throttle_trim, float pitch_limit_min, float pitch_limit_max, float target_climbrate,
+		      float target_sinkrate)
 {
 	// Init subclasses
```
Here the input sanitization is missing. Seems i forgot to at them. Don't simply set the _airspeed_filter to the airspeed value, if it is NaN or zero. Also we need to think of all the other values.

KonradRudin replied:
You did it in TECSAirspeedFilter::initialize which is fine as well

sfuhrer replied:
yes let's assume them sane.

KonradRudin reviewed (changes requested):
nice to make things a bit more robust. Added some remarks, some are just for discussion.

KonradRudin commented:
> @KonradRudin in [f57ae5f](https://github.com/PX4/PX4-Autopilot/commit/f57ae5fbd4d21d581759b105736bec1249e1a201) and following I propose to completely remove the manual TECS resetting from the fw_position_controller, and instead do the initialization only at one place anymore: internally in the `TECS::update()` when the dt is too large. This removes a lot of (seemingly) unnecessary complexity without a downside that I found so far.

In general i think it would be better if the initialization is done outside (and inside only if we know that the timestamp would be unsafe due to CPU spikes or whatever), because it is easier for the executor to know, if it switches to TECS control loop or not. Internally we can only check this based on the timestamp. However, currently the fixedwingpositioncontrol has wierd logic on when to initialize it and it is convoluted as well. So your fix makes it cleaner

KonradRudin reviewed (commented):
Looks good to me. Have you tested this again in SITL with your changes to reproduce it? Also, can we assume that altitude and altitude_rate are always sane? (i guess so).

sfuhrer commented:
Squashed commits and rebased on main.

>Looks good to me. Have you tested this again in SITL with your changes to reproduce it? 

Yes I've tested it again, and it initialized properly now in all my tests (with and without airspeed, on ground and in air).

> Also, can we assume that altitude and altitude_rate are always sane? (i guess so).

For now I would assume them to be sane. Otherwise we'd need to sanitize them on many other places. EKF2 hopefully invalidates the position estimate before it outputs garbage data.
Feel free to to propose additional sanitizations though where ever you deem necessary!

sfuhrer commented:
> I almost crashed because of the underlying issue yesterday when I switched to RTL for a test

Sorry to hear, but happy that you had some altitude margin!

MaEtUgR commented on src/lib/tecs/TECS.cpp:
```diff
@@ -353,7 +360,7 @@ TECSControl::SpecificEnergyRates TECSControl::_calcSpecificEnergyRates(const Alt
 {
-	if (!flag.detect_underspeed_enabled) {
+	if (!flag.detect_underspeed_enabled || !flag.airspeed_enabled) {
 		_ratio_undersped = 0.0f;
```
```suggestion
		_ratio_underspeed = 0.0f;
```

sfuhrer replied:
lets safe that for later

MaEtUgR reviewed (approved):
Thank you for fixing this one 🙏

MaEtUgR commented:
@sfuhrer JFYI I did some testing yesterday and RTL worked nicely exactly like expected in the exact same use case.
Video: https://youtu.be/Ssb4f8lfED8
Log: https://review.px4.io/plot_app?log=d3bd219a-664e-4428-9d94-002824303284
✅

junwoo091400 commented:
Seems like this wasn't added to v1.14 Backport, or did it get added? @dagar
